source/common/protobuf/utility.cc

```diff
@@ -505,6 +506,79 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+// Recursive helper method for MessageUtil::redact() below.
+void redactInPlace(Protobuf::Message* message, bool ancestor_is_redacted) {
+  // If the message is an `Any`, we can't use any type-based introspection. Instead, we have to
+  // unpack it to its original type to redact it...
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    // 1. Extract the type name from the URL. This is guaranteed to be the portion of the URL
+    // following the last '/' character.
+    const std::string& type_url = any->type_url();
+    const std::size_t last_slash = type_url.find_last_of('/');
+    const std::string type_name = type_url.substr(last_slash + 1);
+    const auto* descriptor =
+        Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(type_name);
+
+    // 2. Instantiate a mutable message of the correct dynamic type, and unpack the `Any` into it.
+    Protobuf::DynamicMessageFactory dmf;
+    std::unique_ptr<Protobuf::Message> typed_message(dmf.GetPrototype(descriptor)->New());
+    any->UnpackTo(typed_message.get());
+
+    // 3. Redact the typed message and repack it into the original `Any`.
+    redactInPlace(typed_message.get(), ancestor_is_redacted);
+    any->PackFrom(*typed_message);
+
+    return;
+  }
+
+  // Otherwise, use reflection to traverse all populated fields of this message...
+  const auto* reflection = message->GetReflection();
+  std::vector<const Protobuf::FieldDescriptor*> field_descriptors;
+  reflection->ListFields(*message, &field_descriptors);
+  for (const auto* field_descriptor : field_descriptors) {
+    // Redact if this field or any of its ancestors have the `redacted` option set.
+    const bool redacted =
+        ancestor_is_redacted || field_descriptor->options().GetExtension(envoy::protobuf::redacted);
+
+    if (redacted && field_descriptor->type() == Protobuf::FieldDescriptor::TYPE_STRING) {
+      // Base case: replace strings with "[redacted]".
+      if (field_descriptor->is_repeated()) {
+        const int field_size = reflection->FieldSize(*message, field_descriptor);
+        for (int i = 0; i < field_size; ++i) {
+          reflection->SetRepeatedString(message, field_descriptor, i, "[redacted]");
+        }
+      } else {
+        reflection->SetString(message, field_descriptor, "[redacted]");
+      }
+    } else if (field_descriptor->type() == Protobuf::FieldDescriptor::TYPE_MESSAGE) {
```

Comment:
Will this just ignore redactions of other types, e.g. integer or bool? I feel in general, we should just clear the field.

Reply:
Yes, it just ignores other types. The trouble with clearing the field, as I think Piotr pointed out, is that you can't distinguish between a redacted value and a missing value. Since I'm not aware of a use case yet for redacting non-string values, I decided to punt.

Reply:
I think if the annotation is present, we should either throw an exception or clear to default values, for the other types. Otherwise, people might get a surprise one day, when their secrete integers leak.